### ISSUE TITLE
Container running with "--rm" is not cleaned up after kill -9. Unable to start new containers. Docker ps output is wrong.

### ISSUE THREAD

**Issue by robert-blankenship** [docker/docker#22425]
**Output of `docker version`:**

```
[root@ip-10-10-2-31 ~]# docker --version
Docker version 1.11.0, build 4dc5990
```

**Output of `docker info`:**

```
[root@ip-10-10-2-31 ~]# docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.11.0
Storage Driver: devicemapper
 Pool Name: docker-202:2-27805291-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: (redacted)
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 1.495 GB
 Data Space Total: 107.4 GB
 Data Space Available: 28.23 GB
 Metadata Space Used: 2.109 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.145 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2015-10-14)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay bridge null host
Kernel Version: 3.10.0-327.3.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 14.53 GiB
Name: (redacted)
ID: (redacted)
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Cluster store: (redacted)
Cluster advertise: (redacted)
```

**Steps to reproduce the issue:**
```
# STEP 1: Run container named `hello-github` using `--rm` flag
[root@centos ~]# docker run --rm --name hello-github centos tail -f /etc/hosts`

# STEP 2: Get PID for container.
[root@centos ~]# ps aux | grep "docker run"
root      6776  0.0  0.1 198004 17612 pts/0    Sl+  19:22   0:00 docker run --rm --name hello-github centos tail -f /etc/hosts
root      6984  0.0  0.0 112644   956 pts/1    S+   19:25   0:00 grep --color=auto docker run

# STEP 3: Stop container using kill -9
[root@centos ~]# kill -9 6776

# STEP 4: Start container again:
[root@centos ~]# docker run --rm --name hello-github centos tail -f /etc/hosts
docker: Error response from daemon: Conflict. The name "/hello-github" is already in use by container f893deffaaa43563fd6c6bf007800e1fbbfc831334894511415e7712c387784e. You have to remove (or rename) that container to be able to reuse that name..
See 'docker run --help'
```

**Describe the results you received:**
### Docker ps output is incorrect, because container is not running
![screen shot 2016-04-29 at 12 32 09 pm](https://cloud.githubusercontent.com/assets/7118222/14927401/73b056d4-0e06-11e6-8c7c-98f6d69e6552.png)

### Cannot start new container with same name
![screen shot 2016-04-29 at 12 30 02 pm](https://cloud.githubusercontent.com/assets/7118222/14927404/76f59b56-0e06-11e6-8cb5-4da99a62decd.png)

**Additional steps**
```
# Get IDs of 'docker-containerd' and 'docker-containerd-shim' process
[root@ip-10-10-2-31 ~]# ps aux | grep docker
root      5490  1.1  0.2 1290528 40828 ?       Ssl  19:14   0:14 /usr/bin/docker daemon -H fd:// -H tcp://0.0.0.0:2376 --cluster-store=consul://10.10.2.31:8080 --cluster-advertise=10.10.2.31:2376 --tlsverify --tlscacert=/etc/docker/ssl/exa-ca.pem --tlscert=/etc/docker/ssl/docker-cert.pem --tlskey=/etc/docker/ssl/docker-key.pem --cluster-store-opt kv.cacertfile=/etc/docker/ssl/exa-ca.pem --cluster-store-opt kv.certfile=/etc/docker/ssl/docker-cert.pem --cluster-store-opt kv.keyfile=/etc/docker/ssl/docker-key.pem
root      6915  0.0  0.0 200684  6256 ?        Sl   19:23   0:00 docker-containerd-shim f893deffaaa43563fd6c6bf007800e1fbbfc831334894511415e7712c387784e /var/run/docker/libcontainerd/f893deffaaa43563fd6c6bf007800e1fbbfc831334894511415e7712c387784e docker-runc
root      7040  0.1  0.0 277912  8484 ?        Ssl  19:34   0:00 docker-containerd -l /var/run/docker/libcontainerd/docker-containerd.sock --runtime docker-runc
root      7067  0.0  0.0 112644   952 pts/0    S+   19:34   0:00 grep --color=auto docker

# kill both processes
kill -9 6915 7040

# Confirm that processes are dead; still can't start new Docker container until I manually `docker rm` the container
[root@ip-10-10-2-31 ~]# ps aux | grep docker
root      5490  1.1  0.2 1290528 41080 ?       Ssl  19:14   0:14 /usr/bin/docker daemon -H fd:// -H tcp://0.0.0.0:2376 --cluster-store=consul://10.10.2.31:8080 --cluster-advertise=10.10.2.31:2376 --tlsverify --tlscacert=/etc/docker/ssl/exa-ca.pem --tlscert=/etc/docker/ssl/docker-cert.pem --tlskey=/etc/docker/ssl/docker-key.pem --cluster-store-opt kv.cacertfile=/etc/docker/ssl/exa-ca.pem --cluster-store-opt kv.certfile=/etc/docker/ssl/docker-cert.pem --cluster-store-opt kv.keyfile=/etc/docker/ssl/docker-key.pem
root      7085  0.0  0.0 112644   952 pts/0    S+   19:35   0:00 grep --color=auto docker
```

**Describe the results you expected:**
I expected to be able to start a new container. I don't care too much about the incorrect 'docker ps' output.

**Additional information you deem important (e.g. issue happens only occasionally):**
This situation can arise when the system has a sudden power failure.

**Comment by thaJeztah** [comment]
This is not a bug, because the container is actually still running, and only the docker *client* that's attached to the container is killed;

start the container, and (because it's and interactive session (`-it`) attach to it);

```bash
docker run --rm --name hello-github ubuntu tail -f /etc/hosts
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
172.17.0.4	e9ea043c90ef
```

In another shell, grep for "tail", you'll see *two* processes, one is the docker client session that's attached, the other is the actual container process;

```bash
ps aux | grep "tail"
root      3677  1.0  2.3 191088 24088 pts/3    Sl+  15:56   0:00 docker run --rm --name hello-github ubuntu tail -f /etc/hosts
root      3710  0.4  0.0   4388   836 ?        Ss   15:56   0:00 tail -f /etc/hosts
```

kill the client session, this will attach it from the container;

```bash
$ kill -9 3677
```

but the container is still running;

```bash
ps aux | grep "tail"
root      3710  0.0  0.0   4388   836 ?        Ss   15:56   0:00 tail -f /etc/hosts
```

and you can re-attach to it to continue your session;

```bash
docker attach hello-github

```

However, removing the container (using `-f`, because it's still running);

```bash
$ docker rm -f hello-github
hello-github
```

actually stops the container, and removes it:

```bash
ps aux | grep "tail"
```

**Comment by thaJeztah** [comment]
note that the `--rm` functionality is currently a client-side feature; if the client keeps running, but loses connection with the container, it will trigger a `docker rm`. There's various reasons why it can lose the connection (e.g. network interruption) and in those cases, it cannot connect to the daemon to actually *remove* the container.

There's currently work in progress to improve that, and move the `--rm` to the daemon, so that even if the client gets killed, the daemon can remove the container (directly, or in a later garbage-collect round). You can track progress of that in the PR; https://github.com/docker/docker/pull/20848. There's also other issues linked from the PR description.

I'll close this issue, because this is not a bug, but feel free to comment after I closed

**Comment by robert-blankenship** [comment]
> note that the --rm functionality is currently a client-side feature; if the client keeps running, but loses connection with the container, it will trigger a docker rm. There's various reasons why it can lose the connection (e.g. network interruption) and in those cases, it cannot connect to the daemon to actually remove the container.

That makes sense! Thanks @thaJeztah . We will just add additional logic in our application to ensure that the name is available in the namespace before reinitializing the `hello-github` container.
